File: vscodeExt/209.minimum-size-subarray-sum.py
#
# @lc app=leetcode id=209 lang=python3
#
# [209] Minimum Size Subarray Sum
#

# @lc code=start
class Solution:
    def minSubArrayLen(self, s: int, nums: List[int]) -> int:
        minlen = len(nums) + 1 # 取正常不能获得的最大值
        # if minlen == 0:
        #     return 0
        # if sum(nums) < s:
        #     return 0
        l = 0 
        r = -1 #[l...r] 为滑动区间，不包括任何值
        mySum = 0 
        while r < len(nums):
            if mySum < s :
                r += 1
                if r < len(nums):
                    mySum += nums[r]
            if mySum >= s:
                # print(mySum)
                if r-l + 1 < minlen:
                    minlen = r -l + 1
                mySum -= nums[l]     
                l += 1
        if minlen == len(nums) + 1:
            return 0

        return minlen
# @lc code=end

File: vscodeExt/209.minimum-size-subarray-sum.java
/*
 * @lc app=leetcode id=209 lang=java
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        int left = 0;
        int right = -1;
        int minLength = nums.length+1;
        int windowSum = 0;
        while(right<nums.length){
            if (windowSum < s){
                right +=1;
                // out of Bounds
                if(right<nums.length){
                    windowSum += nums[right];
                }
            } 
            if (windowSum >= s)
            {
                if(right-left < minLength){
                    minLength = right - left;
                }
                windowSum -= nums[left];
                left += 1;
            }
        }
        if(minLength == nums.length +1){
            return 0;
        }

        return minLength +1 ;
    }
}
// @lc code=end

